### ISSUE TITLE
PHP runtime support

### ISSUE THREAD

**Issue by hicksy** [architect/architect#1226]
Mentioned that I'd worked up PHP support for arc in the discord a couple of weeks back. Creating this issue to keep track of things / status and will file some PRs for initial feedback. 

*Sandox*
- [x] Create a php runtime
- [ ] Add PHP to test suite

*Hydrate*
- [x] Hydration through the [Composer](https://getcomposer.org) PHP package manager. 
- [x] Shared / Views deps 
- [ ] Consider a mechanism to remove composer's AWS SDK prior to deploy?
- [ ] Add PHP to test suite

*Inventory*
- [x] Add `.php` as a valid rumtime
- [ ] Update test suite

*Package*
- [x] Add custom layer to functions with PHP runtime
- [ ] Update layer ARN to point to Architect hosted layer
- [ ] Test suite

*Create*
- [x] PHP templates 

**Custom layer**
Current version I've built to PHP 7.4, but I'm going to re-build using PHP 8.0 to [maximise support](https://www.php.net/supported-versions.php)

The WIP layer is currently over here - https://github.com/hicksy/arc-php-layer, and is deployed to[ Serverless Application Repo](https://serverlessrepo.aws.amazon.com/applications/us-east-1/275960034841/arc-php). 

Aside from re-compiling to PHP 8.0, there's a couple of improvements to make to the runtime - most notably to handle binary responses. I'm also compiling with AWS sdk support, but I'll remove that as currently the PHP hydration won't remove the local AWS sdk before deploy 

- [ ] Move to and architect repo
- [ ] Rename / move to architect AWS serverless application repo org
- [ ] Consider automated build / deploy to Serverless Application repo?
- [ ] Keep track with minor version bumps on PHP 8.0
- [ ] Modify layer runtime to handle binary responses

**Architect functions-php**

Will make a start at creating this library and update here with progress

**Comment by ryanblock** [comment]
This is super interesting. An important thing to note: PHP is not an official Lambda runtime: https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html

Thus, relying on PHP means relying on a (3rd party?) PHP custom runtime, and it is not a goal of Architect to support custom runtimes in such a fashion. (Which PHP custom runtime, and why that one over any of the other ones? What if the author of the custom runtime changes the way the handler bootstrap works? What if they un-publish it?)

I think this makes way more sense as an Arc plugin; the plugin API is nascent but this would be an excellent use case for exercising it. wdyt?

**Comment by hicksy** [comment]
@ryanblock Yeah as you say, PHP isn't currently an AWS managed runtime. My thinking was that this custom runtime layer would transfer to Architect ownership and it would therefore become your own managed custom runtime to keep in step with arc, periodic re-builds following PHP 8.0.x updates.

The custom runtime layer is built from from a Dockerfile (ported from [Bref](https://github.com/brefphp/bref)), and the runtime handler is a modified version of this from AWS (https://github.com/aws-samples/php-examples-for-aws-lambda/tree/master/0.7-PHP-Lambda-functions-with-Docker-container-images), and then packed and distributed to SAR.  (Basically followed the same mechanism you've used for Deno.) 

But... I guess you're saying you don't want to manage your own runtime, is that right? 

Plugin API would be interesting - although admittedly, I'm not up to speed with that (it got announced while I was moving house), but I'm keen to dig into it. 

Does the plugin API work across the whole of arc, then? Like... local dev would still be supported that way, and hydration etc?

**Comment by ryanblock** [comment]
(Some context: I added that comment before I noticed that part of what you were doing was including a custom PHP runtime; but I left the comment as-is because the broader questions remained the same!)

Anywho, none of us are active PHP developers, so no, we wouldn't be inclined to volunteer Arc maintainers / contributors to own a custom PHP runtime. And frankly that wouldn't be good for PHP devs either, because we'd surely not do as good of a job as they (you) would.

Nor would I be especially inclined to take this over from the community side and bake it in first-party for all the obvious OSS reasons – if you should happen to stop contributing to Architect, decide to move to the woods and give up software dev, etc. – now we have a core part of the system that no core maintainers use falling into disrepair. So that's a bit tricky. But a community-maintained plugin sounds right on.

Re. the plugin API, it's still in active development, see more here: https://github.com/architect/architect/issues/1175 But tldr yes, the thinking is this would cut across the entire Architect lifecycle.

**Comment by hicksy** [comment]
Ok thanks for the feedback then @ryan

It’s a shame, I’d come away more hopeful from the discord chat, but completely understand where you’re coming from :+1:

If AWS offered PHP as a managed runtime in the future, would that change your opinion about 1st class integration? (I guess I’m asking; is it maintaining a custom layer that’s the main concern, or more widely PHP as a language?)

I’ll reflect on other ways to achieve what we’re looking for, and will definitely keep an eye on how your plugin api comes along - could be super interesting producing a community plugin. PHP coupled with arc’s local dev environment is extremely compelling, and unique to other serverless PHP offerings.

Will revoke the PRs too so they’re not cluttering the place up!

Thanks again

**Comment by ryanblock** [comment]
If AWS offered PHP as an officially supported language we'd definitely consider adding built-in support, but as with all things AWS, the devil is in the details. We'd need to understand how it works and what the implications and impacts would be for Architect, and how much community desire there was for us to add that support.

That said, I am fully supportive of this effort in general. I think it's a great idea and I don't think there's anything fundamentally wrong with the approach (PHP as SAR layer) or goals, which is why I'd love to work with you to help turn this into a plugin.

The plugin API is, as mentioned, nascent, but I think this is precisely the kind of use case we need to help flesh it out and make it better. If you'll bear with us and spend some time collaborating together, I'd love to help support this project's development in that way.

This is the place to start: https://arc.codes/docs/en/guides/extend/plugins Also happy to take this conversation to Discord if you like!

**Comment by hicksy** [comment]
Alright, cool, let's do it 😄

**Comment by ryanblock** [comment]
@hicksy just wanted to provide a followup that I'm currently at work on the next big Arc plugin API release and all your work here is deeply informing it. So first, thank you for putting all this together, we are going to make it happen.

I'm wondering if you'd be able to help test the RCs? Also, any thinking about why not use someone else's PHP custom runtime layer? Example: https://github.com/stackery/php-lambda-layer

**Comment by hicksy** [comment]
@ryanblock oh fantastic - yeah I'd be very happy to help test. 

Someone else's runtime layer would make complete sense, although the one linked is PHP 7.0 / 7.1 - I'd suggest defaulting to one that supports PHP 8.0 as a minimum (latest PHP is 8.1)

The layer I built originally to mock this was based on Bref's docker file (https://github.com/brefphp/bref), with a simplified runtime handler. Then packaged up for AWS Serverless Application Repo - mainly due to originally thinking this would be a core arc runtime rather than a plugin, so just wanted one simplified ARN to ref. 

Now that it's going to be handled by a plugin, I'd imagine a plugin could default to using one of the pre-built / supported layers that the likes of Bref produce, while also giving the dev the ability to ref their own layer ARN if required.
